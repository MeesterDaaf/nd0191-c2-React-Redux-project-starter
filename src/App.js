import NotFound from './components/NotFound';
import Navigation from './components/Navigation';
import { Routes, Route } from 'react-router-dom';
import RequireAuth from './components/RequireAuth';
import PollPage from './components/PollPage';

function App() {
    return (
        <div className="container">
            <Navigation />
            <Routes>
                <Route path="/questions/:id" element={
                    <RequireAuth>
                        {({ question }) =>
                            question ? <PollPage /> : <NotFound />
                        }
                    </RequireAuth>
                } />
                <Route path="*" element={<NotFound />} />

            </Routes>
        </div>
    );
}

export default App; 