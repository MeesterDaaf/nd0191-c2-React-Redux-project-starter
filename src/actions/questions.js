import { _getQuestions, _saveQuestionAnswer } from '../utils/_DATA';

export const RECEIVE_QUESTIONS = 'RECEIVE_QUESTIONS';
export const SAVE_QUESTION_ANSWER = 'SAVE_QUESTION_ANSWER';

export function receiveQuestions(questions) {
    return {
        type: RECEIVE_QUESTIONS,
        questions,
    };
}

export function handleInitialQuestions() {
    return (dispatch) => {
        return _getQuestions().then((questions) => {
            dispatch(receiveQuestions(questions));
        });
    };
}

export function saveQuestionAnswer(authedUser, qid, answer) {
    return {
        type: SAVE_QUESTION_ANSWER,
        authedUser,
        qid,
        answer
    };
}

export function handleSaveAnswer(authedUser, qid, answer) {
    return async (dispatch) => {
        try {
            await _saveQuestionAnswer({
                authedUser,
                qid,
                answer
            });
            dispatch(saveQuestionAnswer(authedUser, qid, answer));
            // Also dispatch user answer update
            dispatch({
                type: 'SAVE_USER_ANSWER',
                authedUser,
                qid,
                answer
            });
        } catch (error) {
            console.error('Error saving answer:', error);
            alert('There was an error saving your answer. Please try again.');
        }
    };
} 