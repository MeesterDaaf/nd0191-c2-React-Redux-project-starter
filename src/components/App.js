import React from 'react';
import { Routes, Route, Navigate } from 'react-router-dom';
import Login from './Login';
import Dashboard from './Dashboard';

const App = () => {
    return (
        <Routes>
            <Route path="/login" element={<Login />} />
            <Route path="/dashboard" element={<Dashboard />} />
            <Route path="/" element={<Navigate to="/login" replace />} />
        </Routes>
    );
}

export default App;