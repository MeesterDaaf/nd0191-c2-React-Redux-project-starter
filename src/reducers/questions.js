import { RECEIVE_QUESTIONS } from '../actions/questions';

export default function questions(state = {}, action) {
    switch (action.type) {
        case RECEIVE_QUESTIONS:
            console.log('Questions reducer received:', action.questions);
            return {
                ...state,
                ...action.questions
            };
        default:
            return state;
    }
} 