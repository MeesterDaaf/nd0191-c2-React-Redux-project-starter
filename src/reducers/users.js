import { RECEIVE_USERS } from '../actions/users';
import { SAVE_USER_ANSWER } from '../actions/users';

export default function users(state = {}, action) {
    switch (action.type) {
        case RECEIVE_USERS:
            return {
                ...state,
                ...action.users
            };
        case SAVE_USER_ANSWER:
            const { authedUser, qid, answer } = action;
            return {
                ...state,
                [authedUser]: {
                    ...state[authedUser],
                    answers: {
                        ...state[authedUser].answers,
                        [qid]: answer
                    }
                }
            };
        case 'ADD_USER_QUESTION':
            return {
                ...state,
                [action.authedUser]: {
                    ...state[action.authedUser],
                    questions: state[action.authedUser].questions.concat([action.qid])
                }
            };
        default:
            return state;
    }
} 