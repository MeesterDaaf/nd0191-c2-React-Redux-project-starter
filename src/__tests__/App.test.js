import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { configureStore } from '@reduxjs/toolkit';
import App from '../components/App';
import authedUser from '../reducers/authedUser';
import users from '../reducers/users';
import questions from '../reducers/questions';

// Create a mock store
const store = configureStore({
    reducer: {
        authedUser,
        users,
        questions
    }
});

describe('App', () => {
    it('should match snapshot', () => {
        const { container } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <App />
                </BrowserRouter>
            </Provider>
        );
        
        expect(container).toMatchSnapshot();
    });
}); 