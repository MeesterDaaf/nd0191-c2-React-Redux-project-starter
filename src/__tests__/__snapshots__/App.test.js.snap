// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <div
    class="flex justify-center items-center px-4 py-12 min-h-screen bg-gray-50 sm:px-6 lg:px-8"
  >
    <div
      class="space-y-8 w-full max-w-md"
    >
      <div>
        <h2
          class="text-3xl font-bold text-center"
        >
          Employee Polls
        </h2>
        <p
          class="mt-2 text-sm text-center text-gray-600"
        >
          Please sign in to continue
        </p>
      </div>
      <form
        class="mt-8 space-y-6"
      >
        <div
          class="rounded-md shadow-sm"
        >
          <select
            class="block relative px-3 py-2 w-full placeholder-gray-500 text-gray-900 rounded-md border border-gray-300 appearance-none focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 focus:z-10 sm:text-sm"
            required=""
          >
            <option
              value=""
            >
              Select a user
            </option>
          </select>
        </div>
        <button
          class="px-4 py-2 w-full text-sm font-medium text-white bg-indigo-600 rounded-md border border-transparent shadow-sm hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
          disabled=""
          type="submit"
        >
          Sign in
        </button>
      </form>
    </div>
  </div>
</div>
`;
